### PR TITLE
Set translation domain for the form validators

### PR DESCRIPTION
**Pull request by c-Rolland**
## Subject

I am targeting this branch, to give the possibility to set a translation domain for the form validators.

## Changelog

```markdown

### Changed
- Add this buildViolation()->setTranslationDomain() in ErrorElement::addViolation
- Add new parameter "$translationDomain" to ErrorElement::addViolation method
```

### CODE REVIEW

File: src/Form/Validator/ErrorElement.php [hunk 1]
```diff
@@ -186,13 +187,15 @@ public function addViolation($message, $parameters = [], $value = null)
         }
 
         $subPath = (string) $this->getCurrentPropertyPath();
+        $domain = (null !== $translationDomain) ? $translationDomain : 'validators';
```
Comment (kunicmarko20) [inline src/Form/Validator/ErrorElement.php]:
```suggestion
        $translationDomain = $translationDomain ?? 'validators';
```

Comment (kunicmarko20) [reply src/Form/Validator/ErrorElement.php]: also I would move `validators` into a `const`

File: src/Form/Validator/ErrorElement.php [hunk 2]
```diff
@@ -186,13 +187,15 @@ public function addViolation($message, $parameters = [], $value = null)
...
 
         if ($this->context instanceof LegacyExecutionContextInterface) {
             $this->context->addViolationAt($subPath, $message, $parameters, $value);
         } else {
             $this->context->buildViolation($message)
                ->atPath($subPath)
                ->setParameters($parameters)
+               ->setTranslationDomain($domain)
```
Comment (kunicmarko20) [inline src/Form/Validator/ErrorElement.php]:
```suggestion
               ->setTranslationDomain($translationDomain)
```

**Comment by core23** [comment]
💯 bonus points if you add a test

File: src/Form/Validator/ErrorElement.php [hunk 3]
```diff
@@ -22,6 +22,8 @@
 
 class ErrorElement
 {
+    protected const DEFAULT_TRANSLATION_DOMAIN = 'validators';
```
Comment (greg0ire) [inline src/Form/Validator/ErrorElement.php]:
```suggestion
    private const DEFAULT_TRANSLATION_DOMAIN = 'validators';
```

File: src/Form/Validator/ErrorElement.php [hunk 4]
```diff
@@ -174,10 +176,11 @@ public function getSubject()
      * @param string|array $message
      * @param array        $parameters
      * @param null         $value
+     * @param string       $translationDomain
      *
      * @return ErrorElement
      */
-    public function addViolation($message, $parameters = [], $value = null)
+    public function addViolation($message, $parameters = [], $value = null, $translationDomain = null)
```
Comment (greg0ire) [inline src/Form/Validator/ErrorElement.php]: This is a BC-break for extending classes.

Comment (c-Rolland) [reply src/Form/Validator/ErrorElement.php]: If i add a new public method who give the possibility to set a translation domain is better ?

Comment (c-Rolland) [reply src/Form/Validator/ErrorElement.php]: if not in a is possible to apply the change in a new version ?

Comment (core23) [reply src/Form/Validator/ErrorElement.php]: https://3v4l.org/mk3eT

You can only do this kind of change on a major version.

Comment (greg0ire) [reply src/Form/Validator/ErrorElement.php]: Or you could simulate a new argument and get it thanks to `func_get_args()`

Comment (greg0ire) [reply src/Form/Validator/ErrorElement.php]: https://3v4l.org/InHLB

File: src/Form/Validator/ErrorElement.php [hunk 5]
```diff
@@ -186,13 +188,22 @@ public function addViolation($message, $parameters = [], $value = null)
         }
 
         $subPath = (string) $this->getCurrentPropertyPath();
+        $translationDomain = self::DEFAULT_TRANSLATION_DOMAIN;
+
+        if (\func_num_args() >= 4) {
```
Comment (core23) [inline src/Form/Validator/ErrorElement.php]: Shoule we add a PHPDoc for this @greg0ire ?

Comment (greg0ire) [reply src/Form/Validator/ErrorElement.php]: I think we should 👍

File: src/Form/Validator/ErrorElement.php [hunk 6]
```diff
@@ -186,13 +189,22 @@ public function addViolation($message, $parameters = [], $value = null)
...
```
Comment (core23) [inline src/Form/Validator/ErrorElement.php]:
```suggestion
        // NEXT_MAJOR: Remove this hack
        if (\func_num_args() >= 4) {
```

Comment (core23) [reply src/Form/Validator/ErrorElement.php]: Can you please merge this @c-Rolland

We would like to do some major refactorings after your PR

Comment (OskarStark) [reply src/Form/Validator/ErrorElement.php]: @core23 I can also apply the suggestion if Rhiannon’s the only thing left to do

Comment (kunicmarko20) [reply src/Form/Validator/ErrorElement.php]: I tried to apply and can't

Comment (OskarStark) [reply src/Form/Validator/ErrorElement.php]: Oh strange, me too 😌

Comment (c-Rolland) [reply src/Form/Validator/ErrorElement.php]: done ^^

**Comment by qsomazzi** [comment]
@sonata-project/release-managers Hello can you create a tag with this change please ? When you got free time of course

**Comment by kunicmarko20** [comment]
Posted on slack channel, should be soon

**Comment by OskarStark** [comment]
@greg0ire are you in the release managers group?

**Comment by greg0ire** [comment]
Yes

**Comment by greg0ire** [comment]
I'm going to try to produce a changelog that is less technical

**Pull request merged**